Gh 271 GitHub secrets

benj-fletch opened the pull request:
This PR covers functionality described in GH-271

The following functionality has been provided by this new code:

Data source `github_actions_public_key` - retrieving the public key of a repository
Data source `github_actions_secrets` - retrieval of all secrets in a repository
Resource `github_actions_secret` - management of Actions Secrets

See the documentation for details around config for each new item.

Note that this requires #342 to be merged prior to this and so it has been raised as a draft PR

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,150 @@
+				Required:  true,
+				Sensitive: true,
+			},
+			"key_id": {
```
Perhaps it's better to not have these arguments and instead just fetch and use the public key on create?

benj-fletch replied:
It's a good shout. I initially provided these to allow users to determine the key / key id without the need for `github_actions_secrets_public_key` but I think just not including them as you suggest is better.

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,157 @@
+		Delete: resourceGithubActionsSecretDelete,
+
+		Schema: map[string]*schema.Schema{
+			"owner": {
```
Why is the `owner` argument required?  Given you use [`checkOrganization`](https://github.com/terraform-providers/terraform-provider-github/blob/b838aed6391ac450740f023e4a714d6cd449aad3/github/util.go#L16) func, that checks that organization is configured in provider settings, you could get the value from provider config via `meta`. As an example, take a look at [repository webhook](https://github.com/terraform-providers/terraform-provider-github/blob/8dc722aeae3d86b082447e22f34ac69539bc277c/github/resource_github_repository_webhook.go#L102) code.

benj-fletch replied:
I was unaware that I could pull the organisation from the `meta` to be honest!
Thinking about this, would there be any case that you would want to manage a secret that wasn't part of the organisation in the provider? I think it might be possible in the current implementation but should not be encouraged - forcing users to have an individual provider per organisation seems more appropriate?

martinssipenko replied:
Yes, same ideology is used when managing other repository resources, webhooks is one example.

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,157 @@
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s:%s", owner, repo, secretName))
```
If we assume that `owner` comes from provider configuration, then you can't have more than a single owner value and the owner part in Id becomes redundant. Have you considered using two part id consisting of repo name and secret name, and combining them using [`buildTwoPartID`](https://github.com/terraform-providers/terraform-provider-github/blob/b838aed6391ac450740f023e4a714d6cd449aad3/github/util.go#L59-L61) utility func, you can find example [here](https://github.com/terraform-providers/terraform-provider-github/blob/8dc722aeae3d86b082447e22f34ac69539bc277c/github/resource_github_repository_deploy_key.go#L84)?

benj-fletch replied:
I agree. I saw that function but wanted to use all 3 values. If we remove `owner` as per above I will implement this suggestion

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,157 @@
+		d.SetId("")
+		return err
+	}
+
```
How come there are no `d.Set()` method calls? I think the secret information could be stored in state. The `UpdatedAt` value from `github.Secret` could be used to determine if the value has changed in GitHub, for example, if `UpdatedAt` stored in state differs from value you got in `GetSecret` response, it means someone has updated the value outside of Terraform, and we probably want to sync it up (reapply value defined in Terraform).

benj-fletch replied:
I wasn't aware that the `d.Set()` functions were required to pass things to state. My understanding was that the existing `d` values passed in would get persisted.
I fully agree with the use of `UpdatedAt` - would you recommend saving the full state of the `github.Secret` ?

martinssipenko replied:
I think I would store the `repo` and `name` as par of Id, then store `plaintext_value` as it is and lastly store `updated_at` as computed getting the value from create/update operation.

benj-fletch replied:
Sounds like a solid plan - I will also include `created_at` since we have it available and might as well make it available

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,157 @@
+	client := meta.(*Organization).client
+	ctx := context.WithValue(context.Background(), ctxId, d.Id())
+
+	orgName := d.Get("owner").(string)
```
Consider getting `orgName` from provider config via `meta` here.

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,157 @@
+	ctx := context.WithValue(context.Background(), ctxId, d.Id())
+
+	orgName := d.Get("owner").(string)
+	repoName := d.Get("repository").(string)
```
Perhaps it's better to obtain `repoName` and `secretName` by extracting it from `d.GetId()` using [`parseTwoPartID`](https://github.com/terraform-providers/terraform-provider-github/blob/b838aed6391ac450740f023e4a714d6cd449aad3/github/util.go#L49-L56) utility func?

```go
repoName, secretName, err := parseTwoPartID(d.Id(), "repository", "secret_name")
if err != nil {
    return err
}
```

martinssipenko commented:
Can't really figure out from the code, are we always updating the secret on GitHub, even if it has not changed locally?

benj-fletch commented:
I didn't think we are updating every time, though reading this review, I was unaware that the secret data wasn't being saved to state so maybe we are always updating. As discussed further up in the review, I think we should add to the state and this will resolve your concern

martinssipenko commented:
@benj-fletch sweet, if you need any help let me know. I'm happy to review again after you have meade the changes discussed above.

benj-fletch commented:
Thanks for the comments @martinssipenko! I have updated the PR to hopefully address all of them. I'll leave it in draft still until the update to go github v29 has occured as I imagine it will create conflicts to resolve

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,166 @@
+	}
+
+	d.SetId(buildTwoPartID(&repo, &secretName))
+	d.Set("plaintext_value", d.Get("plaintext_value"))
```
Not 100% sure but I think is not required, as it already should exist in state.

benj-fletch replied:
I am also not sure on this point. I left it there such that it was obvious (and definite) it would be set. I don't see any issue with leaving this line in?

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,166 @@
+	secretName := d.Get("secret_name").(string)
+
+	secret, _, err := client.Actions.GetSecret(ctx, owner, repo, secretName)
+	if err != nil || secret.Name != secretName {
```
Why is `|| secret.Name != secretName` part needed here?

benj-fletch replied:
I don't think it is on review!

martinssipenko commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		return err
+	}
+
```
You should set the Id here using `d.SetId(buildTwoPartID(&repo, &secretName))`. Afterwards this func calls `resourceGithubActionsSecretRead` func, which instead should extract repo and secret name from the `d.GetId()` via `parseTwoPartID` func, I've given the example below.

martinssipenko commented:
Can you rebase from master? #342 was merged, but unfortunatley it includes v29.0.3, but we need v29.0.3 for this to work. The question is should we make update to latest in this or seperate PR?

benj-fletch commented on github/data_source_github_actions_secrets.go:
```diff
@@ -0,0 +1,106 @@
+	"log"
+)
+
+func dataSourceGithubActionsSecrets() *schema.Resource {
```
I initially added this as I thought it might be useful in future resources / data sources for GitHub actions. Reviewing the available API endpoints I think I agree that this data is useless in the long run. I will remove this data source.

benj-fletch commented on github/resource_github_actions_secret.go:
```diff
@@ -0,0 +1,166 @@
+	owner := meta.(*Organization).name
+	ctx := context.Background()
+
+	repo := d.Get("repository").(string)
```
Yes, I should have caught this in the previous PR comment resolution! Thanks.

benj-fletch commented:
Hi @martinssipenko, thanks a lot for the re-review. I have rebased onto master and hopefully have addressed all your concerns, are you able to have another look?

N.B. This PR now relies on #369 for the go-github v29.0.3 upgrade and will need rebasing again to master following its merge.

btkostner commented:
@benj-fletch #369 has been merged into master now. This can be rebased and hopefully merged! Thank you for all the work on this!

benj-fletch commented:
Thanks for the prod @btkostner - I have rebased, fixed a couple of conflicts and squashed the commits. This should be ready for review now! 💯

jcudit commented on website/docs/r/actions_secret.html.markdown:
````diff
@@ -0,0 +1,55 @@
+
+```
+$ terraform import github_repository.terraform terraform
+```
````
This needs to be updated or removed

benj-fletch replied:
Good catch, classic copy paste error! Will update

jcudit commented on website/github.erb:
```diff
@@ -13,6 +13,9 @@
           <a href="#">Data Sources</a>
           <ul class="nav nav-visible">
+            <li>
+              <a href="/docs/providers/github/d/actions_public_key.html">github_actions_public_key</a>
```
I always miss this file, thanks.
